crates/holochain_sqlite/src/db.rs

```diff
@@ -201,43 +207,109 @@ pub trait WriteManager<'e> {
     where
         E: From<DatabaseError>,
         F: 'e + FnOnce(Transaction) -> Result<R, E>,
     {
-        let txn = self.transaction().map_err(DatabaseError::from)?;
-        f(txn)
+        Ok(self.conn()?.with_reader(f)?)
     }
+}
 
-    #[cfg(feature = "test_utils")]
-    fn with_reader_test<R, F>(&'e mut self, f: F) -> R
+impl<'e> WriteManager<'e> for DbWrite {
+    fn with_commit<E, R, F>(&'e mut self, f: F) -> Result<R, E>
     where
-        F: 'e + FnOnce(Transaction) -> R,
+        E: From<DatabaseError> + std::fmt::Debug,
+        F: 'e + Clone + FnOnce(&mut Transaction) -> Result<R, E>,
     {
-        self.with_reader(|r| DatabaseResult::Ok(f(r))).unwrap()
+        Ok(self.conn()?.with_commit(f)?)
     }
 }
 
-impl<'e> WriteManager<'e> for PConn {
-    fn with_commit<E, R, F>(&'e mut self, f: F) -> Result<R, E>
+impl<'e> ReadManager<'e> for PConn {
+    #[tracing::instrument(skip(self, f))]
+    fn with_reader<E, R, F>(&'e mut self, f: F) -> Result<R, E>
     where
         E: From<DatabaseError>,
-        F: 'e + FnOnce(&mut Transaction) -> Result<R, E>,
+        F: 'e + FnOnce(Transaction) -> Result<R, E>,
     {
-        let mut txn = self
-            .transaction_with_behavior(TransactionBehavior::Exclusive)
+        tracing::trace!("Attempting to acquire read lock on TXN_MUTEX.");
+        let _g = TXN_MUTEX.read();
+        tracing::trace!("Got read lock on TXN_MUTEX.");
+        let txn = self
+            .transaction_with_behavior(TransactionBehavior::Deferred)
             .map_err(DatabaseError::from)?;
-        let result = f(&mut txn)?;
-        txn.commit().map_err(DatabaseError::from)?;
-        Ok(result)
+        let result = f(txn);
+        tracing::trace!("Released read lock on TXN_MUTEX.");
+        result
+    }
+}
+
+impl<'e> WriteManager<'e> for PConn {
+    #[tracing::instrument(skip(self, f))]
+    fn with_commit<E, R, F>(&'e mut self, f: F) -> Result<R, E>
+    where
+        E: From<DatabaseError> + std::fmt::Debug,
+        F: 'e + Clone + FnOnce(&mut Transaction) -> Result<R, E>,
+    {
+        let backoff = backoff::ExponentialBackoff {
+            initial_interval: Duration::from_millis(5),
+            multiplier: 1.5,
+            randomization_factor: 0.1,
+            // TODO: reduce these timeouts significantly after doing some initial
+            // testing to find out how long of a timeout, if any,
+            // fixes our database locking issues.
+            max_interval: Duration::from_secs(1),
+            max_elapsed_time: Some(Duration::from_secs(5)),
+            ..Default::default()
+        };
+
+        // NB: The `?` here converts all errors into `backoff::Error::Transient`.
+        //     If there are unrecoverable errors, we can explicitly map those
+        //     to `backoff::Error::Permanent`.
+        let attempt = || {
+            tracing::trace!("Attempting to acquire write lock on TXN_MUTEX.");
+            let _g = TXN_MUTEX.write();
```

Comment:
You can't use a static because that makes all writes to all databases serial. We want this per database.

Reply:
Oh yeah of ourse